source/agora/common/MerkleTree.d

```diff
@@ -0,0 +1,233 @@
+/*******************************************************************************
+
+    Contains MerkleTree that makes binary tree of transactions.
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.node.MerkleTree;
+
+import agora.common.Data;
+import agora.common.Hash;
+import agora.common.Transaction;
+
+import std.algorithm;
+
+
+/// Ditto
+public struct MerkleTree
+{
+    /// The root of `MerkleTree`
+    public MerkleNode* root;
+
+    /***************************************************************************
+
+        Calculate the hash value of the `MerkleTree`
+
+        Return:
+            The `Hash` representing this instance
+
+    ***************************************************************************/
+
+    public Hash getHash ()
+        nothrow @nogc @trusted
+    {
+        if (this.root is null)
+            return Hash.init;
+        else
+            return this.root.getHash();
+    }
+}
+
+/// The `MerkleNode` has one transaction or two `MerkleNode`
+public struct MerkleNode
+{
+    /// The left branch's ptr of the tree.
+    public MerkleNode* left;
+
+    /// The right branch's ptr of the tree.
+    public MerkleNode* right;
+
+    /// The transaction's ptr
+    public Transaction* tx;
+
+    /// The hash of this `MerkleNode`
+    public Hash hash;
+
+    /***************************************************************************
+
+        Create
+
+        Params:
+            left = The left branch's ptr of the tree.
+            right = The right branch's ptr of the tree.
+            tx = The transaction's ptr
+
+    ***************************************************************************/
+
+    /// Ctor
+    this (MerkleNode* left, MerkleNode* right, Transaction* tx = null)
+    {
+        this.left = left;
+        this.right = right;
+        this.tx = tx;
+        this.hash = this.getHash();
```

Comment:
`getHash` is essentially doing what `computeHash` is doing.
You might want to do:
```suggestion
        this.hash = hashFull(this);
```

Instead, remove `getHash`, and define `computeHash` as:
```D
    {
        // Our node contains either a transaction, in which case it is a leaf node,
        // or 2 childs (which might be the same if the tree is unbalanced)
        if (this.transaction !is null)
            hashPart(*this.transaction, dg);
        else
        {
            hashPart(this.left.hash, dg);
            hashPart(this.right.hash, dg);
        }
}
```